File: src/lib/converter/wow-model/direct/m2/metadata.ts
/**
 * Direct port of M2Exporter's writeMeta: builds the `.json`-equivalent
 * metadata object in memory. Values mirror the legacy JSON exactly (the
 * caller JSON-round-trips the result to normalize undefined/NaN/BigInt the
 * same way JSONWriter + JSON.parse did).
 */
import { getFileNameByID } from '@/lib/wow/archive/client/name-client';
import type { GeosetMaskEntry, TextureManifestEntry, VariantTexture } from '@/lib/wow/export/m2/m2-exporter';
import { M2Loader } from '@/lib/wow/formats/m2/m2-loader';
import { Skin } from '@/lib/wow/formats/m2/skin';

export async function buildMetadataObject(
  m2: M2Loader,
  skin: Skin,
  fileDataID: number,
  fileName: string | undefined,
  geosetMask: GeosetMaskEntry[] | undefined,
  validTextures: Map<VariantTexture, TextureManifestEntry>,
  dataTextures: Set<number>,
): Promise<Record<string, unknown>> {
  // Clone the submesh array and add the 'enabled' property.
  const subMeshes = new Array<unknown>(skin.subMeshes.length);
  for (let i = 0, n = subMeshes.length; i < n; i++) {
    const subMeshEnabled = !geosetMask || geosetMask[i].checked;
    subMeshes[i] = { enabled: subMeshEnabled, ...skin.subMeshes[i] };
  }

  // Clone M2 textures array and expand entries with internal/external names.
  const textures = new Array<unknown>(m2.textures.length);
  for (let i = 0, n = textures.length; i < n; i++) {
    const texture = m2.textures[i];
    const texType = m2.textureTypes[i];
    let textureEntry = validTextures.get(texture.fileDataID);

    // Fallback for data textures whose fileDataID may be 0 or not mapped yet.
    if (!textureEntry && dataTextures.has(texType)) textureEntry = validTextures.get(`data-${texType}`);

    textures[i] = {
      fileNameInternal: typeof texture.fileDataID === 'number' ? await getFileNameByID(texture.fileDataID) : undefined,
      fileNameExternal: textureEntry?.matPathRelative,
      mtlName: textureEntry?.matName,
      ...texture,
    };
  }

  return {
    fileType: 'm2',
    fileDataID,
    fileName,
    internalName: m2.name,
    textures,
    textureTypes: m2.textureTypes,
    materials: m2.materials,
    textureCombos: m2.textureCombos,
    skeletonFileID: m2.skeletonFileID,
    boneFileIDs: m2.boneFileIDs,
    animFileIDs: m2.animFileIDs,
    m2Animations: m2.animations,
    colors: m2.colors,
    textureWeights: m2.textureWeights,
    transparencyLookup: m2.transparencyLookup,
    textureTransforms: m2.textureTransforms,
    textureTransformsLookup: m2.textureTransformsLookup,
    boundingBox: m2.boundingBox,
    boundingSphereRadius: m2.boundingSphereRadius,
    collisionBox: m2.collisionBox,
    collisionSphereRadius: m2.collisionSphereRadius,
    lights: m2.lights || [],
    skin: {
      subMeshes,
      textureUnits: skin.textureUnits,
      fileName: skin.fileName,
      fileDataID: skin.fileDataID,
    },
    cameras: m2.cameras || [],
    cameraLookup: m2.cameraLookup || [],
    ribbonEmitters: m2.ribbonEmitters || [],
    particleEmitters: m2.particleEmitters || [],
  };
}

File: golang/internal/converter/wowmodel/direct/m2/metadata.go
package directm2

import (
	"context"
	"encoding/json"

	m2export "github.com/pqhuy98/wow-converter/internal/wow/export/m2"
	"github.com/pqhuy98/wow-converter/internal/wow/formats/m2"
)

// BuildMetadataObject builds the .json-equivalent metadata object.
func BuildMetadataObject(ctx context.Context, loader *m2.Loader, skin *m2.Skin, fileDataID int, fileName string, geosetMask []m2export.GeosetMaskEntry, validTextures map[any]m2export.TextureManifestEntry, dataTextures map[int]struct{}, getName func(context.Context, int) (string, error)) (map[string]any, error) {
	subMeshes := make([]map[string]any, len(skin.SubMeshes))
	for i, sm := range skin.SubMeshes {
		enabled := geosetMask == nil || (i < len(geosetMask) && geosetMask[i].Checked)
		subMeshes[i] = map[string]any{
			"enabled": enabled, "submeshID": sm.SubmeshID, "level": sm.Level,
			"vertexStart": sm.VertexStart, "vertexCount": sm.VertexCount,
			"triangleStart": sm.TriangleStart, "triangleCount": sm.TriangleCount,
			"boneCount": sm.BoneCount, "boneStart": sm.BoneStart,
			"boneInfluences": sm.BoneInfluences, "centerBoneIndex": sm.CenterBoneIndex,
			"centerPosition": sm.CenterPosition, "sortCenterPosition": sm.SortCenterPosition,
			"sortRadius": sm.SortRadius,
		}
	}

	textures := make([]map[string]any, len(loader.Textures))
	for i, tex := range loader.Textures {
		texType := loader.TextureTypes[i]
		var entry m2export.TextureManifestEntry
		if e, ok := validTextures[tex.FileDataID]; ok {
			entry = e
		} else if tex.FileDataID == 0 && tex.FileName != "" {
			entry, _ = validTextures[tex.FileName]
		} else if dataTextures != nil {
			if _, ok := dataTextures[int(texType)]; ok {
				entry, _ = validTextures["data-"+itoa(int(texType))]
			}
		}
		var internalName any
		if tex.FileDataID > 0 {
			if n, err := getName(ctx, int(tex.FileDataID)); err == nil {
				internalName = n
			}
		} else if tex.FileName != "" {
			internalName = tex.FileName
		}
		textures[i] = map[string]any{
			"fileNameInternal": internalName,
			"fileNameExternal": entry.MatPathRelative,
			"mtlName":          entry.MatName,
			"flags":            tex.Flags,
			"fileDataID":       tex.FileDataID,
		}
	}

	materials := make([]map[string]any, len(loader.Materials))
	for i, mat := range loader.Materials {
		materials[i] = map[string]any{"flags": mat.Flags, "blendingMode": mat.BlendingMode}
	}

	return map[string]any{
		"fileType": "m2", "fileDataID": fileDataID, "fileName": fileName,
		"internalName": loader.Name, "textures": textures, "textureTypes": loader.TextureTypes,
		"materials": materials, "textureCombos": loader.TextureCombos,
		"skeletonFileID": loader.SkeletonFileID, "boneFileIDs": loader.BoneFileIDs,
		"m2Animations": loader.Animations, "textureTransforms": loader.TextureTransforms,
		"textureTransformsLookup": loader.TextureTransformsLookup,
		"transparencyLookup":      loader.TransparencyLookup,
		"boundingBox":             loader.BoundingBox, "boundingSphereRadius": loader.BoundingSphereRadius,
		"collisionBox": loader.CollisionBox, "collisionSphereRadius": loader.CollisionSphereRadius,
		"lights": loader.Lights, "cameras": loader.Cameras, "cameraLookup": loader.CameraLookup,
		"ribbonEmitters": loader.RibbonEmitters, "particleEmitters": loader.ParticleEmitters,
		"colors": loaderColorsJSON(loader.Colors),
		"skin": map[string]any{
			"subMeshes": subMeshes, "textureUnits": skinTextureUnitsJSON(skin.TextureUnits),
			"fileName": skin.FileName, "fileDataID": skin.FileDataID,
		},
	}, nil
}

func skinTextureUnitsJSON(textureUnits []m2.SkinTextureUnit) []map[string]any {
	out := make([]map[string]any, len(textureUnits))
	for i, tu := range textureUnits {
		out[i] = map[string]any{
			"flags":                      tu.Flags,
			"priority":                   tu.Priority,
			"shaderID":                   tu.ShaderID,
			"skinSectionIndex":           tu.SkinSectionIndex,
			"flags2":                     tu.Flags2,
			"colorIndex":                 tu.ColorIndex,
			"materialIndex":              tu.MaterialIndex,
			"materialLayer":              tu.MaterialLayer,
			"textureCount":               tu.TextureCount,
			"textureComboIndex":          tu.TextureComboIndex,
			"textureCoordComboIndex":     tu.TextureCoordComboIndex,
			"textureWeightComboIndex":    tu.TextureWeightComboIndex,
			"textureTransformComboIndex": tu.TextureTransformComboIndex,
		}
	}
	return out
}

func loaderColorsJSON(colors []m2.ColorEntry) []map[string]any {
	if len(colors) == 0 {
		return nil
	}
	b, err := json.Marshal(colors)
	if err != nil {
		return nil
	}
	var out []map[string]any
	if err := json.Unmarshal(b, &out); err != nil {
		return nil
	}
	return out
}
